src/store.rs

```diff
@@ -1271,92 +1124,48 @@ fn for_each_file(
 /// directory is deleted.
 fn cleanup_dir_tree(
     base: &Path,
-    mut keep: impl FnMut(&Path) -> Result<bool, io::Error>
+    mut keep: impl FnMut(&Path) -> Result<bool, Failed>
 ) -> Result<(), Failed> {
     fn recurse(
         base: &Path,
-        op: &mut impl FnMut(&Path) -> Result<bool, io::Error>
+        top: bool,
+        op: &mut impl FnMut(&Path) -> Result<bool, Failed>
     ) -> Result<bool, Failed> {
-        let dir = match fs::read_dir(base) {
-            Ok(dir) => dir,
-            Err(err) if err.kind() == io::ErrorKind::NotFound => {
-                return Ok(false)
-            }
-            Err(err) => {
-                error!(
-                    "Failed to read directory {}: {}",
-                    base.display(), err
-                );
-                return Err(Failed)
+        let dir = if top {
+            match fatal::read_existing_dir(base)? {
+                Some(dir) => dir,
+                None => return Ok(false),
```

Comment:
This is done early because it only applies to the initial directory `cleanup_dir_tree` was called with. I have added a comment pointing this out. (Allowing the directory not to exist really is the only difference between the first and all subsequently recursion steps, so doing this via a flag seemed a good idea.)